bftengine/src/communication/TlsTCPCommunication.cpp

```diff
@@ -245,680 +408,790 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
-                                                   << remPeer);
+        LOG_ERROR(_logger, "peers doesnt match, expected: " << expectedPeerId
+                                                            << ", received: "
+                                                            << remPeer);
         return false;
       }
     }
 
+    // the actual pinning - read the correct certificate from the disk and
+    // compare it to the received one
     namespace fs = boost::filesystem;
     auto path = fs::path(_certificatesRootFolder) / to_string(remotePeerId)
         / connectionType
         / string(connectionType + ".cert");
 
     FILE *fp = fopen(path.c_str(), "r");
     if (!fp) {
-      LOG_ERROR("certificate file not found, path: " << path);
+      LOG_ERROR(_logger, "certificate file not found, path: " << path);
       return false;
     }
 
     X509 *localCert = PEM_read_X509(fp, NULL, NULL, NULL);
     if (!localCert) {
-      LOG_ERROR("cannot parse certificate, path: " << path);
+      LOG_ERROR(_logger, "cannot parse certificate, path: " << path);
       fclose(fp);
       return false;
     }
 
+    // this is actual comparison, compares hash of 2 certs
     int res = X509_cmp(receivedCert, localCert);
+    if (res == 0) {
+      if (_destId == AsyncTlsConnection::UNKNOWN_NODE_ID) {
+        LOG_INFO(_logger,
+                 "connection authenticated, type: " << _connType
+                                                    << ", peer: "
+                                                    << remotePeerId);
+      }
+
+      if(_destId == UNKNOWN_NODE_ID) {
+        _destId = remotePeerId;
+      }
+      _destIsReplica = check_replica(remotePeerId);
+    }
 
     X509_free(localCert);
     fclose(fp);
 
     return res == 0;
   }
+  /// ****************** TLS related functions end ******************** ///
 
-  void close_socket() {
-    LOG_TRACE("enter, node " << _selfId << ", dest: " << _destId
-                             << ", connected: " << connected << ", closed: "
-                             << _closed);
+  /// ************ connect functions ************************** ///
 
-    try {
-      boost::system::error_code ignored_ec;
-      get_socket().shutdown(boost::asio::ip::tcp::socket::shutdown_both,
-                            ignored_ec);
-      get_socket().close();
-      if (_pSslContext)
-        _pSslContext.release();
-    } catch (std::exception &e) {
-      LOG_ERROR("exception, node " << _selfId << ", dest: " << _destId
-                                   << ", connected: " << connected
-                                   << ", ex: " << e.what());
-    }
-
-    LOG_TRACE("exit, node " << _selfId << ", dest: " << _destId
-                            << ", connected: " << connected << ", closed: "
-                            << _closed);
+  void setTimeOut() {
+    _currentTimeout = _currentTimeout == _maxTimeout
+                      ? _minTimeout
```

Comment:
Shouldn't this be `_maxTimeout`? I'm guessing the intention was to cap the time at max, right?